src/lib/Occurrence.hs

```diff
@@ -0,0 +1,623 @@
+  foldl plus zero $ map (iterate injective ib) uses
+
+iterateIxExpr :: NameBinder AtomNameC iter iter'
+              -> IxExpr iter' -> State Bool (IxExpr iter)
+iterateIxExpr ib (Var i) =
+  case hoist ib i of
+    HoistSuccess i' -> return $ Var i'
+    HoistFailure _ ->
+      -- If we hit a reference to the iteration binder,
+      -- that constitutes a proof that indexing is injective,
+      -- because all constructors of `IxExpr` that this could
+      -- be nested under represent invertible functions.
+      -- If we were already injective, this case is an
+      -- over-approximation: the object is actually accessed at just
+      -- one value of the index, but since we don't know which one, we
+      -- assume it was all of them.
+      put True >> return IxAll
+iterateIxExpr ib (Product elts) =
+  -- Rectangular approximation to something like xs.(j, j)
+  Product <$> mapM (iterateIxExpr ib) elts
+iterateIxExpr ib (Inject i elt) =
+  Inject i <$> iterateIxExpr ib elt
+iterateIxExpr ib (Unknown names) =
+  case hoist ib (ListE names) of
+    HoistSuccess (ListE names') -> return $ Unknown names'
+    -- If we hit a reference to an unknown function of the binder, we
+    -- must assume it can hit any data index (which we
+    -- over-approximate as it hitting all of them); and we also do not
+    -- gain a proof that the indexing is injective, because different
+    -- values of the iteration-space binder could lead to the same
+    -- value of the data-space binder.
+    HoistFailure _ -> return IxAll
+iterateIxExpr _ IxAll = return IxAll
+
+-- === CollapsedUse ===
+
+-- CollapsedUse is like Use, but now it is also independent of binders
+-- in scope outside the inlining decision under consideration.
+-- Collapsing Use into CollapsedUse perforce over-approximates those
+-- binders, potentially severely.
+
+-- A CollapsedUse is either a count of the accesses to the whole
+-- object, or a subset of the leading dimension of the object.  In the
+-- latter case, the object is not accessed outside that subset.  The
+-- Subset structure maintains a map of further subsets, and how later
+-- dimensions are accessed in each of them.
+data CollapsedUse =
+    CConst Count
+  | CSubset (Subset CollapsedUse)
+  deriving (Eq, Show)
+
+-- A subset of an index type.  The constructors correspond to possible
+-- index types---each one gives known information about the structure
+-- of the type, and relevant available subset information.
+-- A trie on subtypes of an index type that supports subset
+-- intersection as efficiently as is practicable.
+-- We include an optimization: The Inj constructor carries an optional
+-- payload associated with the whole type, instead of broadcasting it
+-- over all the cases.
+data Subset a =
+  -- The index could be any type, and all of it is accessed.
+    SAll a
+  -- The index is a sum.  We maintain a map from constructors to
+  -- subsets of their respective types.
+  | Inj  (Maybe a) (M.IntMap (Subset a))
```

Comment:
What is the `Maybe`?

Reply:
It's an optional value that's associated with all the cases, as an optimization over explicitly broadcasting it.  Tried to clarify the comment.